### PR TITLE
RADIUS Management User Authentication Feature

### PR DESCRIPTION
**Pull request by a-barboza**
#### What I did

Radius Management User Authentication Feature

#### How I did it

HLD: https://github.com/Azure/SONiC/blob/master/doc/aaa/radius_authentication.md

#### How to verify it

This is the CLI only. The changes are reflected in the Redis Config DB.

#### New command output (if the output of a command-line utility has changed)

admin@sonic:~$ show radius
RADIUS global auth_type pap (default)
RADIUS global retransmit 3 (default)
RADIUS global timeout 5 (default)
RADIUS global passkey <EMPTY_STRING> (default)

admin@sonic:~$

admin@sonic:~$ sudo config radius
Usage: config radius [OPTIONS] COMMAND [ARGS]...

  RADIUS server configuration

Options:
  -?, -h, --help  Show this message and exit.

Commands:
  add         Specify a RADIUS server
  authtype    Specify RADIUS server global auth_type [chap | pap | mschapv2]
  default     set its default configuration
  delete      Delete a RADIUS server
  nasip       Specify RADIUS server global NAS-IP|IPV6-Address <IPAddress>
  passkey     Specify RADIUS server global passkey <STRING>
  retransmit  Specify RADIUS server global retry attempts <0 - 10>
  sourceip    Specify RADIUS server global source ip <IPAddress>
  statistics  Specify RADIUS server global statistics [enable | disable |...
  timeout     Specify RADIUS server global timeout <1 - 60>
admin@sonic:~$

### CODE REVIEW

**Comment by lguohan** [comment]
need unit test to be added.

**Comment by a-barboza** [comment]
> need unit test to be added.

Done.

**Comment by a-barboza** [comment]
UT's are added and seemed to have passed.
...
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_show_radius_default PASSED        [ 81%]
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_config_radius_server PASSED       [ 82%]
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_config_radius_server_invalidkey PASSED [ 82%]
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_config_radius_nasip_invalid PASSED [ 82%]
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_config_radius_sourceip_invalid PASSED [ 82%]
[2021-03-25T22:41:02.610Z] tests/radius_test.py::TestRadius::test_config_radius_authtype PASSED     [ 82%]
...
build completes
...
[2021-03-25T22:47:41.168Z] Successfully tagged docker-sonic-vs:sonic-utilities-build-pr.4574
[2021-03-25T22:47:41.168Z] + docker save docker-sonic-vs:sonic-utilities-build-pr.4574
...
sonic-swss-tests, fail in an unrelated test because of a unrelated error.
...
[2021-03-25T22:48:57.568Z] + git clone https://github.com/Azure/sonic-swss sonic-swss-tests
[2021-03-25T22:48:57.568Z] Cloning into 'sonic-swss-tests'...
...
[2021-03-26T01:31:24.475Z] ==================================== ERRORS ====================================
[2021-03-26T01:31:24.475Z] _______________ ERROR at setup of TestVxlan.test_vxlan_term_orch _______________
[2021-03-26T01:31:24.475Z] 
[2021-03-26T01:31:24.475Z] self = <docker.transport.unixconn.UnixHTTPConnectionPool object at 0x7facdcc26d30>
[2021-03-26T01:31:24.475Z] conn = <docker.transport.unixconn.UnixHTTPConnection object at 0x7facdcc7abb0>
...

**Comment by a-barboza** [comment]
The system lag timing issues in the build tests are unrelated. They appear to have been fixed here (https://github.com/Azure/sonic-swss/pull/1692)

File: config/aaa.py [hunk 1]
```diff
@@ -61,20 +71,69 @@ def fallback(option):
...
 
     if 'default' in auth_protocol:
+        if len(auth_protocol) !=1:
+            click.echo('Not a valid command')
+            return
         del_table_key('AAA', 'authentication', 'login')
     else:
         val = auth_protocol[0]
         if len(auth_protocol) == 2:
-            val += ',' + auth_protocol[1]
```
Comment (renukamanavalan) [inline config/aaa.py:77]: a choice *only*:
s = sorted(l)
if s[0] == "local" and s[1] in ("radius", "tacacs+")

File: config/aaa.py [hunk 2]
```diff
@@ -189,3 +248,249 @@ def delete(address):
...
+    """set its default configuration"""
+    ctx.obj = 'default'
+radius.add_command(default)
+
+
+@click.command()
+@click.argument('second', metavar='<time_second>', type=click.IntRange(1, 60), required=False)
+@click.pass_context
+def timeout(ctx, second):
+    """Specify RADIUS server global timeout <1 - 60>"""
```
Comment (renukamanavalan) [inline config/aaa.py:271]: The command is "config radius timeout".
Please correct the comments

File: tests/radius_test.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+config_radius_invalidipaddress_output="""\
+Invalid ip address
+"""
+
+class TestRadius(object):
+    @classmethod
+    def setup_class(cls):
+        os.environ['UTILITIES_UNIT_TESTING'] = "1"
+        print("SETUP")
+        import config.main
```
Comment (renukamanavalan) [inline tests/radius_test.py:44]: does this check o/p of "show aaa" with login as radius

File: config/aaa.py (hunk 2 above)
Comment (a-barboza) [reply config/aaa.py:271]: The comment string is the help string in the CLI output. I've followed the same style as TACACS+. Which comment should I correct?

`admin@sonic:~$ sudo config  radius timeout --help
Usage: config radius timeout [OPTIONS] <time_second>

  Specify RADIUS server global timeout <1 - 60>

Options:
  -h, -?, --help  Show this message and exit.
admin@sonic:~$`

File: tests/radius_test.py (hunk 3 above)
Comment (a-barboza) [reply tests/radius_test.py:44]: It checks 'show radius' output. I could not find any existing 'show aaa' UT for TACACS+. Does a new test need to be added for 'show aaa' with RADIUS login?

Comment (renukamanavalan) [reply tests/radius_test.py:44]: It will be nice, if it could be added. Can we ?

Comment (a-barboza) [reply tests/radius_test.py:66]: Done.

**Review by renukamanavalan — APPROVED** [review]
Thank you, Arun!

**Pull request merged**